proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,1189 @@
+.. author:: Matthew Pickering, Rodrigo Mesquita, Adam Gundry
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/682>`_.
+.. contents::
+.. sectnum::
+
+
+Explicit Level Imports
+======================
+
+Template Haskell allows *staged metaprogramming*: writing Haskell code that is
+executed during the compilation stage, rather than merely compiled so that it
+can be executed at the later runtime stage. In practice, most programs can be
+divided into small portions that are executed at compile-time (e.g. the
+definition of ``makeLenses``), and the majority of the code that is not.
+However, there are currently no clear boundaries between stages, so the compiler
+must pessimistically assume that anything may be needed at compile-time or
+runtime.
+
+This has a variety of negative consequences. In particular:
+
+* Using Template Haskell causes compile-time performance to suffer due to
+  unnecessary (re)compilation.  This is particularly relevant for interactive
+  use of the compiler within an IDE (e.g. via Haskell Language Server).
+
+* Cross-compilation needs to compile and execute code on the target platform
+  during the build process, as there is no way to execute splices on the host.
+
+*Levels* are the mechanism which the typechecker uses to ensure that staged evaluation
+is possible. Certain program contexts require identifiers to be at specific levels.
+In order to fix level-incorrect programs, programmers rely on implicit *cross-stage persistence* (CSP): the assumption that if
+a definition can be used at one level, then it can also be used at a later level. Cross-stage persistence is the root
+cause of the unnecessary compilation: if a module is used at compile time (e.g. as a dependency
+of a module with ``TemplateHaskell`` enabled),
+the compiler must assume it may be needed for runtime as well.
+
+Thus we propose a new
+pair of extensions that allow programmers to write level-correct programs
+without ubiquitous cross-stage persistence:
+
+* ``NoImplicitStagePersistence`` forbids normal top-level identifiers from
```

Comment:
One of the things that I've learned from #601 and #669 is that framing extensions in terms of a negative like this ends up being confusing. At least for a couple of situations/discussions/users. Would it be possible to restate this without the leading `No`?

Reply:
This is similar to e.g. [`FieldSelectors`](https://downloads.haskell.org/ghc/latest/docs/users_guide/exts/field_selectors.html#extension-FieldSelectors), or [`StarIsType`](https://downloads.haskell.org/ghc/latest/docs/users_guide/exts/poly_kinds.html#extension-StarIsType). The positive `ImplicitStagePersistence` would be included in the existing language editions.

Reply:
To amend my previous comment. The proposal says 

> The ImplicitStagePersistence extension is introduced to control the existing path-based cross stage persistence behaviour. This can now be disabled to force programmers to control levels specifically with staged imports.

We could try to phrase that differently starting from no implicit stage persistence state, and saying that

> The `ImplicitStagePersistence` extension can be (and will be by default) turned on to free programmers from specifying import levels.

Reply:
> To amend my previous comment. The proposal says
> 
> > The ImplicitStagePersistence extension is introduced to control the existing path-based cross stage persistence behaviour. This can now be disabled to force programmers to control levels specifically with staged imports.
>

Thank you! I missed that the `ImplicitStagePersistence` was to be the default. My concern is around the overall presentation, particularly as the number of extensions is already an issue for some.

Reply:
Thanks @phadej and @telser, good points!

>  It's a bit trickier to think which language is (e.g. syntactically) smaller, with or without implicit stage persistence, but arguably one without is "more explicit", i.e. simpler.

I believe enabling implicit stage persistence (alone) should result in strictly more programs being accepted. That's part of the rationale for framing it this way: `NoImplicitStagePersistence` really is taking away a (mis)feature of the language, much as `NoStarIsType` does.

Reply:
Yes @telser, the proposal is to introduce an extension called `ImplicitStagePersistence`, but we talk about it as if the extension is called `NoImplicitStagePersistence` because it's easier to motivate the change from the current default. 

As Oleg points out, in the specification section we specify it in terms of `ImplicitStagePersistence`.

Reply:
> Thanks @phadej and @telser, good points!
> 
> > It's a bit trickier to think which language is (e.g. syntactically) smaller, with or without implicit stage persistence, but arguably one without is "more explicit", i.e. simpler.
> 
> I believe enabling implicit stage persistence (alone) should result in strictly more programs being accepted. That's part of the rationale for framing it this way: `NoImplicitStagePersistence` really is taking away a (mis)feature of the language, much as `NoStarIsType` does.

Now I wonder, what happens if user has `ImplicitStagePersistence` enabled, but explicitly specifies the stages (for some imports). Are they ignored, or are they treated as explicit?

And IIRC I pointed out there is no keyword for default stage, which I'd like to see.

If I understand correctly, you cannot have `ImplicitStagePersistence` and explicit stage keywords: this is bad design in my opinion, the syntax and semantic changes are bundled together (c.f. problems with `Strict`, cannot explicitly use `~` to make lazy fields without enabling `StrictData`: and that's unfortunate).

Reply:
> Now I wonder, what happens if user has `ImplicitStagePersistence` enabled, but explicitly specifies the stages (for some imports). Are they ignored, or are they treated as explicit? (Apparently a parse error?)

* Under `NoExplicitLevelImports + ImplicitStagePersistence` (the default), it is an error to explicitly specify stages.
* Under `ExplicitLevelImports + ImplicitStagePersistence` (which must be specified in that order, because of the extension implication), it is permitted to specify levels but they are effectively ignored (because the implicit cross-stage persistence). This is mentioned briefly at the end of section 1.2.2.

> And IIRC I pointed out there is no keyword for default stage, which I'd like to see.

This is a fair point I think. The biggest challenge I see is figuring out what the keyword should be...

> If I understand correctly, you cannot have `ImplicitStagePersistence` and explicit stage keywords: this is bad design in my opinion, the syntax and semantic changes are bundled together (c.f. problems with `Strict`, cannot explicitly use `~` to make lazy fields without enabling `StrictData`: and that's unfortunate). And good example is `GADTSyntax` and `GADTs`: the syntax and "semantics" are split.

We've learned this from you already, which is why `ImplicitStagePersistence` controls the semantics and `ExplicitLevelImports` the syntax. :-)